crates/codegen/parser/runtime/src/cursor.rs

```diff
@@ -359,62 +359,55 @@ impl Cursor {
         false
     }
 
-    pub fn find_matching<R, F: Fn(&Cursor) -> Option<R>>(&mut self, filter_map: F) -> Option<R> {
-        while !self.is_completed {
-            if let Some(result) = filter_map(self) {
-                return Some(result);
+    /// Attempts to go to the next token, according to the DFS pre-order traversal.
+    ///
+    /// Returns `false` if the cursor is finished and at the root.
+    pub fn go_to_next_token(&mut self) -> bool {
+        while self.go_to_next() {
+            if matches!(&self.current.node, Node::Token(_)) {
+                return true;
             }
-            self.go_to_next();
         }
 
-        None
+        return false;
     }
 
-    /// In contrast to `Iterator::find_*`, this does not consume the first item when found.
-    fn find_noconsume<F: Fn(&Node) -> Option<R>, R>(&mut self, predicate: F) -> Option<R> {
-        while !self.is_completed {
-            match predicate(&self.current.node) {
-                Some(result) => return Some(result),
-                _ => {
-                    self.go_to_next();
-                }
+    /// Attempts to go to the next token with any of the given kinds, according to the DFS pre-order traversal.
+    ///
+    /// Returns `false` if the cursor is finished and at the root.
+    pub fn go_to_next_token_with_kinds(&mut self, kinds: &[TokenKind]) -> bool {
+        while self.go_to_next() {
+            if matches!(&self.current.node, Node::Token(token) if kinds.contains(&token.kind)) {
+                return true;
             }
         }
 
-        None
+        return false;
     }
 
-    /// Finds the first token with either of the given kinds.
+    /// Attempts to go to the next rule, according to the DFS pre-order traversal.
     ///
-    /// Does not consume the iterator if the first item matches.
-    pub fn find_token_with_kind(&mut self, kinds: &[TokenKind]) -> Option<Rc<TokenNode>> {
-        self.find_noconsume(|node| node.as_token_with_kind(kinds).cloned())
-    }
+    /// Returns `false` if the cursor is finished and at the root.
+    pub fn go_to_next_rule(&mut self) -> bool {
+        while self.go_to_next() {
+            if matches!(&self.current.node, Node::Rule(_)) {
+                return true;
+            }
+        }
```

Comment:
I still think it's clearer to extract a helper that does
```rust
while self.go_to_next() {
    if pred(&self.current.node) {
        return true;
     }
}

return false;
```
rather than repeat it 4 times here but I'm not going to block on this

Reply:
Sure.